### DISCUSSION TITLE
Associate a FLAMEGPU_AGENT_FUNCTION to different agents types

### DISCUSSION THREAD

**Discussion by daniele-baccega** [FLAMEGPU/FLAMEGPU2#1072] [Q&A] [answered]
Hi,
is it possible to associate the same FLAMEGPU_AGENT_FUNCTION to different agents types? For example the function that handle the movements. If yes, how can I do it?

Thank you!

**Comment by Robadob** [comment]
I believe you should just be able to attach the same function to multiple agents (assuming they have all the used variables). You would simply need to call `AgentDescription::newFunction()` twice, once for each agent, and then configure settings such as optional message/agent death a second time (you could use the same agent function implementation with different message lists per agent, or twice within the same agent even).

If you are constructing your model via layers, you will need to use a different function name when calling `newFunction()` or pass the `AgentFunctionDescription` handles rather than using the name of the `FLAMEGPU_AGENT_FUNCTION` macro.

Are you getting any particular errors?

**Reply by daniele-baccega** [reply]
I tried using different function name, but I got this error at runtime:
**terminate called after throwing an instance of 'flamegpu::exception::InvalidAgentFunc'
  what():  /home/trustalert/Desktop/cottolengo-hospital-flame-gpu-2/build/_deps/flamegpu2-src/include/flamegpu/model/LayerDescription.h(349): There are 5 possible agent functions to add to layer, please use a more specific method for adding this agent function to a layer, in LayerDescription::addAgentFunction().**

**Reply by Robadob** [reply, marked as answer]
You're recieving that because you're passing the `FLAMEGPU_AGENT_FUNCTION` macro.

e.g. passing `foo` with the below agent function.

```c++
FLAMEGPU_AGENT_FUNCTION(foo, flamegpu::MessageNone, flamegpu::MessageNone) {
   // Do something
}
```

If you have added that agent function to multiple agents, FLAMEGPU won't know which you're referring to (hence "5 possible agent functions").

Instead where you add it to an agent, e.g.

```c++
AgentDescription agent1 = model.newAgent("agent1");
...
AgentFunctionDescription fn1 = agent1.newFunction("foo_fn", foo);
```

You can either add it to layer by `AgentFunctionDescription` instance, or agent and function name (`"agent1"` `"foo"`).

```c++
LayerDescription layer1 = model.newLayer();
layer1.addAgentFunction(fn1);  // This works: Add it by AgentFunctionDescription instance
layer1.addAgentFunction("agent1", "foo_fn");  // This works: Add it by agent and function name
layer1.addAgentFunction(foo);  // This does not work if multiple agents use foo
```

Hope this helps

**Reply by Robadob** [reply]
I've created a PR to clarify the multiple options, with a note regarding multiple use of agent functions to the docs, will hopefully be merged in the near future.

https://github.com/FLAMEGPU/FLAMEGPU2-docs/pull/144

Thanks
